### PR TITLE
Provisioning s3datarepository

### PR DESCRIPTION
**Pull request by tenyo**
This PR adds the core functionality for provisioning an S3 data repository (bucket):
  1. Check if the requested bucket already exists in S3
  2. Create the bucket and wait for it to be successfully created
  3. Block all public access to the bucket
  4. Enable AWS managed serverside encryption (AES-256) for the bucket
  5. Add tags to the bucket

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -10,4 +10,6 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus/client_golang v1.4.1
 	github.com/sirupsen/logrus v1.4.2
+	golang.org/x/text v0.3.2
+	google.golang.org/appengine v1.6.5
```
Comment (fishnix) [inline go.mod]: dont need these probably, they probably got added by your editor when you typo'd something

File: s3datarepository/s3datarepository.go [hunk 2]
```diff
@@ -143,14 +151,188 @@ func WithPrefix(prefix string) S3RepositoryOption {
...
+	return true, nil
+}
+
+// Provision creates and configures a data repository in S3, and creates a default IAM policy
+// 1. Check if the requested bucket already exists in S3
+// 2. Create the bucket and wait for it to be successfully created
+// 3. Block all public access to the bucket
+// 4. Enable AWS managed serverside encryption (AES-256) for the bucket
+// 5. Add tags to the bucket
+func (s *S3Repository) Provision(ctx context.Context, id string, rawtags []*dataset.Tag) error {
```
Comment (fishnix) [inline s3datarepository/s3datarepository.go]: `rawtags` is a weird name

File: s3datarepository/s3datarepository.go [hunk 3]
```diff
@@ -143,14 +151,188 @@ func WithPrefix(prefix string) S3RepositoryOption {
...
+				},
+			},
+		},
+	}); err != nil {
+		return ErrCode("failed to enable encryption for s3 bucket "+name, err)
+	}
+
+	// add tags
+	if len(tags) > 0 {
+		log.Debugf("adding tags for bucket: %s\n%+v", name, tags)
```
Comment (fishnix) [inline s3datarepository/s3datarepository.go]: :eyes: newline in log statement

File: s3datarepository/s3datarepository.go [hunk 4]
```diff
@@ -143,14 +151,188 @@ func WithPrefix(prefix string) S3RepositoryOption {
...
+	}
+
+	log.Debugf("deleting s3datarepository with id: %s", id)
+
+	name := "dataset-" + id
+
+	// delete the s3 bucket
+	_, err := s.S3.DeleteBucketWithContext(ctx, &s3.DeleteBucketInput{
+		Bucket: aws.String(name),
+	})
```
Comment (fishnix) [inline s3datarepository/s3datarepository.go]: this is perfectly fine, but I would probably move it to be one line

File: s3datarepository/s3datarepository_test.go [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+)
+
+// mockS3Client is a fake S3 client
+type mockS3Client struct {
+	s3iface.S3API
+	t         *testing.T
+	err       map[string]error
```
Comment (fishnix) [inline s3datarepository/s3datarepository_test.go]: interesting way to handle mocking more than one call in a tested function, i might steal it 😆 

you could wrap the error too... but this is more straightforward.

File: s3datarepository/s3datarepository_test.go [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+		"region":   "us-east-1",
+		"akid":     "xxxxx",
+		"secret":   "yyyyy",
+		"bucket":   "somethingspecial",
+		"prefix":   "slash",
+		"endpoint": "https://under.mydesk.amazonaws.com",
+	}
+
+	s, err := NewDefaultRepository(testConfig)
+
```
Comment (fishnix) [inline s3datarepository/s3datarepository_test.go]: ✂️

File: s3datarepository/s3datarepository_test.go [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+	}
+}
+
+func (m *mockS3Client) HeadBucketWithContext(ctx context.Context, input *s3.HeadBucketInput, opts ...request.Option) (*s3.HeadBucketOutput, error) {
+	if m.err["HeadBucketWithContext"] != nil {
+		if m.headCount == 0 {
+			m.headCount++
+			return nil, m.err["HeadBucketWithContext"]
+		}
+	}
```
Comment (fishnix) [inline s3datarepository/s3datarepository_test.go]: this isn't wrong, but these should probably use the form:

```golang
if err, ok := m.err["HeadBucketWithContext"]; ok {
....
}
```

when getting the value of a nonexistent key in a map, you always get the zero value of the type... here the zero value of error is `nil`, but it's more confusing IMO than using the “comma ok” idiom.

File: s3datarepository/s3datarepository_test.go [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+	err = s.Provision(context.TODO(), "68004EEC-6044-45C9-91E5-AF836DCD9234", []*dataset.Tag{})
+	if err != nil {
+		t.Errorf("expected nil error, got: %s", err)
+	}
+
+	// test empty id
+	id = ""
+	expectedCode = apierror.ErrBadRequest
+	expectedMessage = "invalid input"
+	s.S3.(*mockS3Client).err = make(map[string]error)
```
Comment (fishnix) [inline s3datarepository/s3datarepository_test.go]: `newMockS3Client` does this already, right?

File: s3datarepository/workflow.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
+package s3datarepository
```
Comment (fishnix) [inline s3datarepository/workflow.go]: yes bikeshed, but `workflow` is not my favorite name 😆

File: go.mod (hunk 1 above)
Comment (tenyo) [reply go.mod]: `go mod tidy` did the trick

**Comment by tenyo** [comment]
I made those fixes and also added the Org as a tag and as a part of the dataset name

File: api/handlers_datasets.go [hunk 10]
```diff
@@ -109,15 +113,15 @@ func (s *server) DatasetCreateHandler(w http.ResponseWriter, r *http.Request) {
 	}()
 
 	// create dataset storage location
-	if err = dataRepo.Provision(r.Context(), id, input.Tags); err != nil {
+	if err = dataRepo.Provision(r.Context(), Org, id, input.Tags); err != nil {
```
[the remark this answers is not in the document]
Comment (tenyo) [reply api/handlers_datasets.go]: I've removed the Org from here and added a `NamePrefix` to the S3Repository struct

File: go.mod (hunk 1 above)
Comment (fishnix) [reply go.mod]: 👍

**Pull request merged**